### DISCUSSION TITLE
Design details of the Babel AST

### DISCUSSION THREAD

**Discussion by magic-akari** [babel/babel#14834] [Q&A] [answered]
I came across https://github.com/swc-project/swc/issues/5416 and want to figure out some details of the Babel AST design.

1. Why is typeAnnotion attached to Pat and not to VariableDeclarator?

```TypeScript
const [a]: [number] = 1;
```

2. Why is the parameter default value parsed as an AssignmentPattern instead of an init field in Param?

```JavaScript
function foo(a = 1) { }
```

**Comment by nicolo-ribaudo** [comment, marked as answer]
Mostly for legacy reasons, I hope to change it: https://github.com/babel/babel/issues/9545

**Comment by JLHwung** [comment]
> Why is the parameter default value parsed as an AssignmentPattern instead of an init field in Param?

So that `function foo(a = 1) {}`, `function ({ foo = 1 }) {}` have consistent ASTs. I guess the name assignment pattern comes from https://tc39.es/ecma262/#prod-AssignmentPattern, though the Babel AssignmentPattern also covers the BindingElement with an initializer.

**Reply by magic-akari** [reply]
No. It's different.

I'm investigating the `typeAnnotion`, so I noticed the AssignmentPattern.
In conclusion, Pattern does not support `typeAnnotion` in all cases.

See this example.
```typescript
function foo({ a: number = 1 }) {}
```
_Note: this is not valid in TypeScript or flow.js._



So the best way to handle it is to remove `typeAnnotion` from Pattern, and hoist it to VariableDeclarator or Param.

But we should support this.

```typescript
function foo(a: number = 1 ) {}
```

Since we are going to remove `typeAnnotion` from Pattern, we cannot parse it as AssignmentPattern anymore.

---
This is the topic discussed in #9545.

**Reply by JLHwung** [reply]
> Pattern does not support typeAnnotion in all cases.

ArrayPattern and ObjectPattern could have type annotations: `({ a }: { a: number }) => {}`. Unlike `swc`, Babel does not have function parameter wrapper node like `Param`, so `{ a }: { a: number }` in this example is an object pattern. In Babel, a functions's param can only be one of the type

https://github.com/babel/babel/blob/e2442e15cae4b3532adcc7a61a8cf1d7f7d23bf6/packages/babel-types/src/ast-types/generated/index.ts#L445

I understand the discrepancy between `function foo({ a: number = 1 }) {}` and `function foo(a: number = 1 ) {}`. And we have similar cases in JS, too:
```js
(a = 1) = b; // invalid, AssignmentPattern can not be LHS
[a = 1] = b; // valid, AssignmentPattern can be nested in LHS
```

we may solve them by the new AST design, but changing function's AST will have huge impact. We should carefully weigh such decisions.

**Reply by magic-akari** [reply]
Thanks for the explanation. I thought we would follow the https://github.com/babel/babel/issues/9545#issuecomment-549514519 way.


If there is no wrapped AST node, there is no place to hoist `typeAnnotion`.
hoisting typeAnnotion and defining new nodes will become the same thing.
